File: code/section_iii/3_4_retrieving_key_for_insert.py
from sqlalchemy import create_engine, text

engine = create_engine('sqlite:///thunderbird_manufacturing.db')
conn = engine.connect()

insertStmt = text("INSERT INTO CUSTOMER (CUSTOMER_NAME, ADDRESS, CITY, STATE, ZIP, CATEGORY) VALUES (:customerName,:address,:city,:state,:zip,:category)")

insertResults = conn.execute(insertStmt, customerName ="Brittania Solutions, LLC",
                             address="2374 Collie Way",
                             city="Frisco",
                             state="TX",
                             zip= 75035,
                             category="COMMERCIAL"
                             )

print("Last inserted record has CUSTOMER_ID {0}".format(insertResults.lastrowid))

conn.close()

File: code/section_iii/3_4_retrieving_key_for_insert.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;

public class JavaLauncher {

    public static void main(String[] args) {

        try {

            //Connection conn = DriverManager.getConnection("jdbc:sqlite:/c:/git/oreilly_intermediate_sql_for_data/thunderbird_manufacturing.db");
            Connection conn = DriverManager.getConnection("jdbc:sqlite:///Users/thomasnield/git/oreilly_intermediate_sql_for_data/thunderbird_manufacturing.db");
            PreparedStatement insertStmt = conn.prepareStatement("INSERT INTO CUSTOMER (CUSTOMER_NAME, ADDRESS, CITY, STATE, ZIP, CATEGORY) VALUES (?,?,?,?,?,?)");

            insertStmt.setString(1, "Brittania Solutions, LLC");
            insertStmt.setString(2, "2374 Collie Way");
            insertStmt.setString(3, "Frisco");
            insertStmt.setString(4, "TX");
            insertStmt.setInt(5, 75035);
            insertStmt.setString(6, "COMMERCIAL");

            // Execute INSERT
            insertStmt.execute();

            // Get the generated key for the new record
            int newRecordId = insertStmt.getGeneratedKeys().getInt(1);

            // Get the new record
            PreparedStatement selectStmt = conn.prepareStatement("SELECT * FROM CUSTOMER WHERE CUSTOMER_ID = ?");
            selectStmt.setInt(1, newRecordId);
            ResultSet rs = selectStmt.executeQuery();

            while (rs.next()) {
                System.out.println(rs.getInt("CUSTOMER_ID") + "," + rs.getString("CUSTOMER_NAME") + "," + rs.getString("STATE") + "," + rs.getString("CATEGORY"));
            }

            conn.close();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}
